### DISCUSSION TITLE
How to debug possible fan incompatibility ?

### DISCUSSION THREAD

**Discussion by 4erk4** [tjko/fanpico#41] [Q&A] [answered]
I'm seeing fans that appear to ignore (or mostly ignore) the `fixed, <PWM>` setting, as measured by the RPM readback, and by visual inspection.

The fan seems to move faster with a jump from 10-100 - but this is subjective. 

The current configuration is just 2 fans,  identical Coolermaster Sickleflow 120.
```
fan1,"fan1",660,22.01,100.0
fan2,"fan2",663,22.11,50.0
fan3,"fan3",0,0.00,100.0
fan4,"fan4",0,0.00,100.0
fan5,"fan5",0,0.00,100.0
fan6,"fan6",0,0.00,100.0
fan7,"fan7",0,0.00,100.0
fan8,"fan8",0,0.00,100.0
````

Fan1 should be at `100.0 PWM` which should translate to 1800 RPM (as per spec)
Fan2 is at `50.0 PWM` which should be about half that.

I am not using the `mbfan` at all. 

Link to specification:
https://www.coolermaster.com/catalog/coolers/case-fan/sickleflow-120/#specifications

I am using the 804D kit, self-assembled, without a display
`TJKO Industries,FANPICO-0804D,<OMITTED>,1.4.1`

The fans section from config seems reasonable:
```
"fans": [{                                                                                                 [168/1502]
"id":   0,
"name": "fan1",
"min_pwm":      0,
"max_pwm":      100,
"pwm_coefficient":      1,
"source_type":  "fixed",
"source_id":    100,
"pwm_map":      [[0, 0], [100, 100]],
"rpm_factor":   2,
"filter":       {
"name": "none"
}
}, {
"id":   1,
"name": "fan2",
"min_pwm":      0,
"max_pwm":      100,
"pwm_coefficient":      1,
"source_type":  "fixed",
"source_id":    50,
"pwm_map":      [[0, 0], [100, 100]],
"rpm_factor":   2,
"filter":       {
"name": "none"
}
```
Any thoughts as to what I can try?

**Comment by tjko** [comment]
Those fans seems to be standard 12V PC PWM (4-pin) fans, so those should just work... 

Are you able to try some other brand/model PWM fans, to see if those behave any differently?


I did quick test on the 1.4.1 firmware and I didn't see any problem setting fixed PWM output for fans, so I don't think its firmware issue.

Can you try increasing logging verbosity maybe to "INFO":  ```sys:log INFO```

Then manually switching PWM frequency and see what gets logged. When I tested it seemed to work fine (on a Noctua and Corsair fans that I had laying around).

```
sys:log info
[272088.948827][0] Change log level: NOTICE (5) -> INFO (6)
conf:fan2:source fixed,50
[272104.908242][0] fan2: change source fixed,100 --> fixed,50
[272105.685004][1] fan2: Set output PWM 100.0% --> 50.0%
[272106.185254][1] fan2: Input Tacho change 58.12Hz --> 56.27Hz
[272106.685201][1] mbfan2: Set output Tacho 58.12Hz --> 56.27Hz
[272107.185257][1] fan2: Input Tacho change 56.27Hz --> 41.98Hz
[272107.685197][1] mbfan2: Set output Tacho 56.27Hz --> 41.98Hz
[272108.185248][1] fan2: Input Tacho change 41.98Hz --> 32.22Hz
[272108.685219][1] mbfan2: Set output Tacho 41.98Hz --> 32.22Hz
```

(change on fan speed should be almost immediate and at least on my test fans there was audible change on noise level)


(If you have access to Oscilloscope it would be interesting to see what you see on the PWM pin of the FAN connector...)

**Reply by 4erk4** [reply]
Thanks for the 'info' tip  (picocom doesn't log commands, so here is going from 100 -> 50 -> 10 -> 100 
```
[ 83558.563185][0] fan2: change source fixed,100 --> fixed,50
[ 83559.339242][1] fan2: Set output PWM 100.0% --> 50.0%

[ 83584.041800][0] fan2: change source fixed,50 --> fixed,10
[ 83585.339475][1] fan2: Set output PWM 50.0% --> 10.0%

[ 83745.998497][0] fan2: change source fixed,10 --> fixed,100
[ 83747.341506][1] fan2: Set output PWM 10.0% --> 100.0%
[ 83747.841774][1] sensor2: Temperature change 33.0C --> 32.5C
```

* I don't currently have access to an Oscilloscope.

I originally had another fan "Arctic P12 Max" - which was much louder because it defaults to the top speed instead of the lowest - the RPM didn't change with the PWM changes either.

**Reply by tjko** [reply]
Just to confirm, these fans have 4-pin connectors and are connected to the "FAN" connectors (and not "MB FAN" connectors)?
How are you powering the fans via the AUX or from the matching MBFAN connector?

Is this the exact same fan as you have? https://www.amazon.com/gp/product/B086RC5BBS/


Btw, you can enable "echo" by using ```sys:echo 1```, that makes it easier to see if you type commands by hand....

**Reply by 4erk4** [reply]
> Just to confirm, these fans have 4-pin connectors and are connected to the "FAN" connectors (and not "MB FAN" connectors)? How are you powering the fans via the AUX or from the matching MBFAN connector?
> 

Confirming that these fans have 4 pin connectors, are connected to FAN (1 and 2). 
MBFAN is not connected, power is through AUX (floppy drive connector). 
Connection is through USB2 (on motherboard) -> serial (haven't enabled wifi yet) 

> Is this the exact same fan as you have? https://www.amazon.com/gp/product/B086RC5BBS/

No, this one
https://www.amazon.com/gp/product/B086HNZWD7/    (The v2 model, black, no RGB)
The specs seem the same though


> Btw, you can enable "echo" by using `sys:echo 1`, that makes it easier to see if you type commands by hand....

Excellent, I had missed that in the commands list

**Reply by tjko** [reply]
I ordered one of those fans (should get it tomorrow), as I'm now curious if there is something different with these fans or is it something else.

Can you double check JP1 connector is set to "AUX" (and not "FAN1" since you are powering via AUX)?  As not having power on 5V power rail could explain why fan speed control is not working, which would happen if powering only via AUX connector and JP1 is set to "FAN1" position.

**Reply by 4erk4** [reply]
Wow.   Thank you!  

I've confirmed that JP1 is set to AUX.   All jumpers are set to AUX, since I wouldn't be using the MBFAN interface. 

Is there a LED or other quick way of checking the 5V rail.  The Pico itself seems to draw power from USB instead of the 5v on aux, and remains powered on the 5v standby rail.

**Reply by tjko** [reply]
LED D4 should be illuminated when there is 5V power (there is test point TP3 next to it to measure the voltage as well).

You should also see "12V AUX" LED (D2) illuminated as well as "3.3V" (D5) illuminated?

**Reply by 4erk4** [reply]
There is a constellation of lights, so I didn't realize there was another one there. D'oh!

LED D4 isn't lit, and the testpoint doesn't seem to have any voltage 
3.3v and 12v aux are both illuminated

`sensors` on Linux doesn't give a 5v rail, but since USB2 works to power, I have 5v.   I haven't measured at the floppy connector yet.

**Reply by tjko** [reply, marked as answer]
If D4 is not lit, that would point the 5V rail being down, and that would explain why fan speed control is not working (as the TTL voltage level converter chips need 5V power).

5V power rail is either powered from the MBFAN1 connector or AUX power connector, JP1 controls which is the source. So you'd want to double check that pins 1 and 2 on JP1 are connected together (with a jumper)...

Btw, you may want to double check all soldering joints for the connectors and headers/jumpers.  I've myself several times missed soldering joint or two when trying to quickly assemble a FanPico...

**Reply by 4erk4** [reply]
*One* weak solder joint.  *One*   
And on the JP1 side .. 
3.3v lights up,  12v lights up.

I'm deeply sorry for causing trouble.   At best, I can be an example to others of what to check for

**Reply by tjko** [reply]
Glad that you got yours working.  I updated Wiki with list of LEDs and test points, and added note about making sure that "5V" LED is lit.

**Comment by tjko** [comment]
If you have an oscilloscope, you should see 25kHz PWM signal on the PWM (4th) pin on the FAN connector.

It should measure 50% duty cycle when PWM output is set to "50%" on the fanpico for that fan port:
![DS1Z_QuickPrint81](https://github.com/tjko/fanpico/assets/1075726/c1e28b52-252e-4dc2-b2e9-3a0a4ed66d9c)


And there should be constant ~5V signal when PWM output is set to "100%":
![DS1Z_QuickPrint82](https://github.com/tjko/fanpico/assets/1075726/dba4493d-a74a-4be7-acb2-70a7b64ed874)


Also, with a digital multimeter you should see about 5V on the PWM pin when output is set to "100%" and close to 0V (in practice you probably see 0.2-0.3V) when output is set to "0%".    Also some multimeters will display about 2.5V when output is set to 50%, but that depends on the meter (how its acquisition/measurement works).
